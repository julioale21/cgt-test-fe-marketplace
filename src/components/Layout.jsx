import { Outlet } from 'react-router-dom';
import { Header } from '../components/shared/Header';
import { Box } from '@mui/material';

export function Layout() {
  const cartItemsCount = 3;

  return (
    <Box sx={{ display: 'flex', flexDirection: 'column', minHeight: '100vh' }}>
      <Header cartItemsCount={cartItemsCount} />

      <Box component="main" sx={{ flexGrow: 1, p: 3 }}>
        <Outlet />
      </Box>

      <Box component="footer" sx={{ p: 2, textAlign: 'center', mt: 'auto' }}>
        <p>Â© 2025</p>
      </Box>
    </Box>
  );
}
