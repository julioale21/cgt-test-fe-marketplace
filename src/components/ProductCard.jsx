import React from 'react';

import { Link } from 'react-router-dom';

import { Button, Card, CardActions, CardContent, CardMedia, Grid, Typography } from '@mui/material';
import ShoppingCartIcon from '@mui/icons-material/ShoppingCart';

const ProductCard = ({ product }) => {
  return (
    <Grid item xs={12} sm={6} md={4} lg={3}>
      <Card sx={{ height: '100%', display: 'flex', flexDirection: 'column' }}>
        <CardMedia
          component="img"
          height="200"
          image={product.image}
          alt={product.name}
          sx={{
            objectFit: 'cover',
            height: '200px'
          }}
        />
        <CardContent sx={{ flexGrow: 1 }}>
          <Typography gutterBottom variant="h6" component="h2">
            {product.name}
          </Typography>
          <Typography variant="body2" color="text.secondary" sx={{ mb: 2 }}>
            {product.description}
          </Typography>
          <Typography variant="h6" color="primary" sx={{ mt: 2 }}>
            ${product.price.toFixed(2)}
          </Typography>
          <Typography variant="body2" color="text.secondary" sx={{ mt: 1 }}>
            Formats: {product.format.join(', ')}
          </Typography>
        </CardContent>
        <CardActions sx={{ justifyContent: 'space-between' }}>
          <Button component={Link} to={`/products/${product.id}`} size="small" color="primary">
            View Details
          </Button>
          <Button size="small" variant="contained" color="primary" startIcon={<ShoppingCartIcon />}>
            Add to Cart
          </Button>
        </CardActions>
      </Card>
    </Grid>
  );
};

export { ProductCard };
