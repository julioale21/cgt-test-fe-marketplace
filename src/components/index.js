export { ContainedButton } from "./buttons/ContainedButton";
export { Header } from "./shared/Header";
export { Layout } from "./Layout";
export { OutlinedButton } from "./buttons/OutlinedButton";
export { ProductList } from "./ProductList";
export {ProductCard} from "./ProductCard";
