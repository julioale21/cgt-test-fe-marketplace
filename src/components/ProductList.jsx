import React from 'react';
import { Grid } from '@mui/material';
import { ProductCard } from './ProductCard';

const ProductList = ({ products }) => {
  return (
    <Grid container spacing={4}>
      {products.map((product) => (
        <ProductCard key={product.id} product={product} />
      ))}
    </Grid>
  );
};

export { ProductList };
