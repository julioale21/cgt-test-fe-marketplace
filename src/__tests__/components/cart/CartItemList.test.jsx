import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { CartItemList } from '../../../components';

jest.mock('@mui/material', () => ({
  ...jest.requireActual('@mui/material'),
  Stack: ({ children, spacing }) => (
    <div data-testid="mui-stack" className="MuiStack-root" data-spacing={spacing}>
      {children}
    </div>
  ),
  Box: ({ children, ...props }) => (
    <div {...props} className="MuiBox-root">
      {children}
    </div>
  )
}));

jest.mock('../../../components/cart/CartItem', () => ({
  CartItem: ({ item, onIncrement, onDecrement, onRemove }) => (
    <div className="MuiPaper-root" data-testid={`cart-item-${item.id}`}>
      <div>
        <span>{item.name}</span>
        <span>${item.price.toFixed(2)}</span>
        <button data-testid={`increment-${item.id}`} onClick={() => onIncrement()}>
          +
        </button>
        <span data-testid={`quantity-${item.id}`}>{item.quantity}</span>
        <button data-testid={`decrement-${item.id}`} onClick={() => onDecrement()}>
          -
        </button>
        <button data-testid={`remove-${item.id}`} onClick={() => onRemove()}>
          x
        </button>
      </div>
    </div>
  )
}));

describe('CartItemList', () => {
  const mockItems = [
    {
      id: 1,
      name: '3D Mountain Model',
      price: 29.99,
      image: '/mountain.jpg',
      quantity: 1
    },
    {
      id: 2,
      name: '3D Tree Model',
      price: 19.99,
      image: '/tree.jpg',
      quantity: 2
    }
  ];

  const defaultProps = {
    items: mockItems,
    onIncrement: jest.fn(),
    onDecrement: jest.fn(),
    onRemove: jest.fn()
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const props = { ...defaultProps };
    const { container } = render(<CartItemList {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correct number of CartItem components', () => {
    const props = { ...defaultProps };
    render(<CartItemList {...props} />);
    const cartItems = screen.getAllByTestId(/^cart-item-/);
    expect(cartItems).toHaveLength(mockItems.length);
  });

  it('should render items in correct order with all required elements', () => {
    const props = { ...defaultProps };
    render(<CartItemList {...props} />);

    const items = screen.getAllByTestId(/^cart-item-/);
    expect(items).toHaveLength(2);
    items.forEach((item, index) => {
      const mockItem = mockItems[index];
      expect(item).toHaveTextContent(mockItem.name);
      expect(item).toHaveTextContent(mockItem.price.toFixed(2));
      expect(screen.getByTestId(`quantity-${mockItem.id}`)).toHaveTextContent(
        mockItem.quantity.toString()
      );
    });
  });

  it('should call onIncrement with correct item', async () => {
    const props = { ...defaultProps };
    const user = userEvent.setup();

    render(<CartItemList {...props} />);
    await user.click(screen.getByTestId('increment-1'));

    expect(props.onIncrement).toHaveBeenCalledTimes(1);
    expect(props.onIncrement).toHaveBeenCalledWith(mockItems[0]);
  });

  it('should call onDecrement with correct item', async () => {
    const props = { ...defaultProps };
    const user = userEvent.setup();

    render(<CartItemList {...props} />);
    await user.click(screen.getByTestId('decrement-2'));

    expect(props.onDecrement).toHaveBeenCalledTimes(1);
    expect(props.onDecrement).toHaveBeenCalledWith(mockItems[1]);
  });

  it('should call onRemove with correct item id', async () => {
    const props = { ...defaultProps };
    const user = userEvent.setup();

    render(<CartItemList {...props} />);
    await user.click(screen.getByTestId('remove-1'));

    expect(props.onRemove).toHaveBeenCalledTimes(1);
    expect(props.onRemove).toHaveBeenCalledWith(mockItems[0].id);
  });

  it('should handle empty items array', () => {
    const props = { ...defaultProps, items: [] };
    render(<CartItemList {...props} />);
    const cartItems = screen.queryAllByTestId(/^cart-item-/);
    expect(cartItems).toHaveLength(0);
  });
});
