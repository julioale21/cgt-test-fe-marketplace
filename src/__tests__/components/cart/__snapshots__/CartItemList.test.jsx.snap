// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItemList should match snapshot 1`] = `
<div>
  <div
    class="MuiStack-root"
    data-spacing="2"
    data-testid="mui-stack"
  >
    <div
      class="MuiPaper-root"
      data-testid="cart-item-1"
    >
      <div>
        <span>
          3D Mountain Model
        </span>
        <span>
          $
          29.99
        </span>
        <button
          data-testid="increment-1"
        >
          +
        </button>
        <span
          data-testid="quantity-1"
        >
          1
        </span>
        <button
          data-testid="decrement-1"
        >
          -
        </button>
        <button
          data-testid="remove-1"
        >
          x
        </button>
      </div>
    </div>
    <div
      class="MuiPaper-root"
      data-testid="cart-item-2"
    >
      <div>
        <span>
          3D Tree Model
        </span>
        <span>
          $
          19.99
        </span>
        <button
          data-testid="increment-2"
        >
          +
        </button>
        <span
          data-testid="quantity-2"
        >
          2
        </span>
        <button
          data-testid="decrement-2"
        >
          -
        </button>
        <button
          data-testid="remove-2"
        >
          x
        </button>
      </div>
    </div>
  </div>
</div>
`;
